src/task/executor.rs

```diff
@@ -70,91 +71,147 @@ impl Clone for Kicker {
     }
 }
 
-struct NotifyContext {
-    kicked: bool,
+const NOTIFIED: u8 = 1;
+const IDLE: u8 = 2;
+const POLLING: u8 = 3;
+const COMPLETED: u8 = 4;
+
+/// A custom notify.
+///
+/// It will poll the inner future directly if it's notified on the
+/// same thread as inner cq.
+pub struct SpawnTask {
+    handle: UnsafeCell<SpawnHandle>,
+    state: AtomicU8,
     kicker: Kicker,
 }
 
-impl NotifyContext {
+impl SpawnTask {
+    fn new(s: Spawn<BoxFuture<(), ()>>, kicker: Kicker) -> SpawnTask {
+        SpawnTask {
+            handle: UnsafeCell::new(Some(s)),
+            state: AtomicU8::new(IDLE),
+            kicker,
+        }
+    }
+
     /// Notify the completion queue.
     ///
     /// It only makes sense to call this function from the thread
     /// that cq is not run on.
-    fn notify(&mut self, tag: Box<CallTag>) {
-        match self.kicker.kick(tag) {
-            // If the queue is shutdown, then the tag will be notified
-            // eventually. So just skip here.
-            Err(Error::QueueShutdown) => (),
-            Err(e) => panic!("unexpected error when canceling call: {:?}", e),
-            _ => (),
+    fn mark_notified(&self) -> bool {
+        loop {
+            match self.state.compare_exchange_weak(
+                IDLE,
+                NOTIFIED,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            ) {
+                Ok(_) => return true,
+                Err(POLLING) => match self.state.compare_exchange_weak(
+                    POLLING,
+                    NOTIFIED,
+                    Ordering::AcqRel,
+                    Ordering::Acquire,
+                ) {
+                    Err(IDLE) | Err(POLLING) => continue,
+                    _ => return false,
+                },
+                Err(IDLE) => continue,
+                _ => return false,
+            }
         }
     }
 }
 
-/// A custom notify.
-///
-/// It will poll the inner future directly if it's notified on the
-/// same thread as inner cq.
-#[derive(Clone)]
-pub struct SpawnNotify {
-    ctx: Arc<SpinLock<NotifyContext>>,
-    handle: Arc<SpinLock<SpawnHandle>>,
-    worker_id: ThreadId,
+pub fn resolve(cq: &CompletionQueue, task: Arc<SpawnTask>, success: bool) {
+    // it should always be canceled for now.
+    assert!(success);
+    poll(cq, task, true);
 }
 
-impl SpawnNotify {
-    fn new(s: Spawn<BoxFuture<(), ()>>, kicker: Kicker, worker_id: ThreadId) -> SpawnNotify {
-        SpawnNotify {
-            worker_id,
-            handle: Arc::new(SpinLock::new(Some(s))),
-            ctx: Arc::new(SpinLock::new(NotifyContext {
-                kicked: false,
-                kicker,
-            })),
+impl Notify for WorkerInfo {
```

Comment:
This implementation is not safe because `WorkerInfo` is not threadsafe.

Reply:
`WorkerInfo` is threadsafe as explained above.